### DISCUSSION TITLE
Indexing _id for Nested/Subdocuments

### DISCUSSION THREAD

**Discussion by AChangXD** [typegoose/typegoose#824] [Q&A] [answered]
So I have a Booking object as such:

```ts
export class Booking {
  @Field(() => String)
  readonly _id: string | mongoose.Types.ObjectId;

  /* ------------------------------ Time metadata ----------------------------- */
  @Field(() => Date)
  @Property({ required: true, index: true })
  startTime: Date;

  @Field(() => Date)
  @Property({ required: true, index: true })
  endTime: Date;

  @Field(() => [User])
  @Property({ type: [User], required: true, default: [], index: true })
  bookableUsers: User[];

  @Field(() => [Bookable])
  @Property({ type: [Bookable], required: true, default: [], index: true })
  bookables: Bookable[];
}
```

Do I have to specify a specific field (_id in this case) for the array subdocuments (bookableUsers  & bookables) in this case? According to the docs: https://www.mongodb.com/docs/manual/core/index-multikey/ -> "MongoDB automatically creates a multikey index if any indexed field is an array; you do not need to explicitly specify the multikey type." If I do need to specify the field, how could I do so?

**Comment by AChangXD** [comment]
@hasezoey Also on another note: syncIndexes() also creates indexes for nested documents. For example: Location has a unique field called name, typegoose will create the index for location.name to be unique within the Booking collection. Is this a bug or desired behavior? Doesn't make sense for Location's unique indecies to be carried over when it's used as a nested document

**Reply by hasezoey** [reply]
> Is this a bug or desired behavior?

from what i know, this is "wanted" behavior from mongoose, the way to "workaround" this is to create a shared base class, then if you dont want indexes you can use the base class directly, and otherwise create a extending class which has the indexes and use that class for the model

or as you have noted in the other comment, use schema-option `excludeIndexes`

**Reply by AChangXD** [reply]
Yeah extending classes just seems to just complicate things, so I went with the excludeindexes option. Can I can still do @Index on the parent to index my subdocuments' _id? If so what is the correct way to do so?

**Reply by hasezoey** [reply]
> Can I can still do https://github.com/Index on the parent to index my subdocuments' _id? If so what is the correct way to do so?

should be able to do it, you can create indexes for lower paths with a `.`, like `parentProp.nestedProp`

**Reply by AChangXD** [reply]
Got it, thank you so much!

**Reply by AChangXD** [reply]
> > Can I can still do https://github.com/Index on the parent to index my subdocuments' _id? If so what is the correct way to do so?
> 
> should be able to do it, you can create indexes for lower paths with a `.`, like `parentProp.nestedProp`

On a related note, when you save a document with a subdocument (the subdocument is also in a collection of it's own), typegoose will auto-populate all other fields in the subdocument. Example being a User collection having:
firstName
lastName
passwordHashed

If passwordHashed has the {select: false} option, it wouldn't be fetched while quering for it. After quering for the user, when I try to save the user to the document, the User's options that wasn't specified (passwordHashed for example) gets auto populated when saving as a standalone subdocument. But when you save the users in an array (array of subdocuments), no issue with auto-populating when savingI

**Reply by AChangXD** [reply]
@hasezoey  ~~set also doesn't seem like they are doing anything on singular sub-documents~~:
`@Property({
    required: true,
    excludeIndexes: true,
    get: (customer: User) => {
      return customer;
    },
    set: (customer: User) => {
      return { ...customer, password: null, resetToken: null };
    },
  })
  customer: User;`

~~Not sure if this is a bug.~~
Nvm it's working, I feel like it's a Janky work-around when the single subdocument needs a getter/setter while the array doesn't

**Comment by AChangXD** [comment]
I think this was brought up in 2020 by @rubenvereecken: https://github.com/Automattic/mongoose/issues/9128 and https://github.com/szokodiakos/typegoose/issues/407 . I can't seem to find the issue in the new repo

**Comment by AChangXD** [comment, marked as answer]
Seems like excludeIndexes: true is the answer!
